Add the first part of the blog post about parsing SQL Server log files (#1166)
debezium/debezium.github.io: _posts/2025-09-05-sqlserver-tx-log.adoc

---
layout: post
title:  Peeking Inside the SQL Server Transaction Log
date:   2025-08-29 00:00:00 -0000
tags: [ sqlserver, tx-log, cdc-tables, internals]
author: vjuranek
---

> SQL Server offers change data capture (CDC) through link:https://learn.microsoft.com/en-us/sql/relational-databases/track-changes/about-change-data-capture-sql-server?view=sql-server-ver17[captured tables] - a special set of system tables that record modifications in selected “ordinary” tables.

Reviewer: I think the correct term is change tables, not captured tables - see https://learn.microsoft.com/en-us/sql/relational-databases/track-changes/about-change-data-capture-sql-server?view=sql-server-ver17#change-table

If you want to monitor changes in real time, you query these captured tables periodically.
That’s exactly how Debezium works today: it polls SQL Server’s captured tables at configured intervals and turns the results into a continuous stream of CDC records.
This approach works fine, but could we do better?

> Captured tables are filled by the SQL Server Agent, which reads the transaction log, extracts changes, and stores them for in captured tables.

Reviewer: ```suggestion
Captured tables are filled by the SQL Server Agent, which reads the transaction log, extracts changes, and stores them in captured tables.
```

In theory, we could skip the middleman and parse the transaction log directly.

> That’s how tools like OpenLogReplicator can handle CDC for Oracle databases.

Reviewer: ```suggestion
That’s how tools like OpenLogReplicator handles CDC for Oracle databases.
```

[source, sql]
----
SELECT file_id, vlf_sequence_number, vlf_begin_offset, vlf_size_mb FROM sys.dm_db_log_info(DB_ID('TestDB'));
----

Example output containing a VLF sequence number (unique per file),  begin offset (where it starts in the file) and size in MB looks like this:

[source]
----
file_id vlf_sequence_number  vlf_begin_offset     vlf_size_mb              
------- -------------------- -------------------- ------------------------
     2                   42                 8192       1.9299999999999999
     2                    0              2039808       1.9299999999999999
     2                    0              4071424       1.9299999999999999
     2                    0              6103040       2.1699999999999999
----

Besides the VLF sequent number, another interesting thing is the VLF offset.
The first VLF actually doesn’t start at the very beginning of the log file, but after 8192 bytes of the file.
The first 8kB are reserved for the log file header.

VLFs header contains information whether VLFs is active or not.
VLFs are used in a circular fashion and unused VLFs can be reused for newly created VLFs.

> VLF cannot be reused until it’s in an active state.

Reviewer: ```suggestion
VLF cannot be reused ehrn it’s in an active state.
```
